src/ws/cockpit-session-client-certificate.c

```diff
@@ -138,69 +153,79 @@ https_instance_has_certificate_file (char   *contents,
   if (https_instance == NULL) /* already warned */
     goto out;
 
+  /* No tricky stuff, please */
+  if (!valid_256_bit_hex_string (filename))
+    {
+      warnx ("tls-cert authentication token is invalid");
+      goto out;
+    }
+
+  if (strcmp (filename, https_instance) != 0)
+    {
+      warnx ("tls-cert authentication token doesn't match wsinstance cgroup");
+      goto out;
+    }
+
   dirfd = open ("/run/cockpit/tls", O_PATH | O_DIRECTORY | O_NOFOLLOW);
   if (dirfd == -1)
     {
       warn ("Failed to open /run/cockpit/tls");
       goto out;
     }
 
-  filefd = openat (dirfd, https_instance, O_RDONLY | O_NOFOLLOW);
+  filefd = openat (dirfd, filename, O_RDONLY | O_NOFOLLOW);
   if (filefd == -1)
     {
-      warn ("Failed to open certificate file /run/cockpit/tls/%s", https_instance);
+      warn ("Failed to open certificate file /run/cockpit/tls/%s", filename);
       goto out;
     }
 
   if (fstat (filefd, &buf) != 0)
     {
-      warn ("Failed to stat certificate file /run/cockpit/tls/%s", https_instance);
+      warn ("Failed to stat certificate file /run/cockpit/tls/%s", filename);
       goto out;
     }
 
   if (!S_ISREG (buf.st_mode))
     {
-      warnx ("Could not read certificate: /run/cockpit/tls/%s is not a regular file", https_instance);
+      warnx ("Could not read certificate: /run/cockpit/tls/%s is not a regular file", filename);
       goto out;
     }
 
   if (buf.st_size == 0)
     {
-      warnx ("Could not read certificate: /run/cockpit/tls/%s is empty", https_instance);
+      warnx ("Could not read certificate: /run/cockpit/tls/%s is empty", filename);
       goto out;
     }
 
-  if (contents != NULL)
```

Comment:
Add an assertion instead?

Reply:
That also changes in future commits.  I don't want to get too caught up on the in-between stuff.